src/lib/x509/certstor_system_windows/certstor_windows.cpp

```diff
@@ -0,0 +1,236 @@
+      Handle_Guard(Handle_Guard<T>&& rhs) :
+         m_context(std::move(rhs.m_context))
+         {
+         rhs.m_context = nullptr;
+         }
+
+      ~Handle_Guard()
+         {
+         close<T>();
+         }
+
+      operator bool() const
+         {
+         return m_context != nullptr;
+         }
+
+      bool assign(T context)
+         {
+         m_context = context;
+         return m_context != nullptr;
+         }
+
+      T& get()
+         {
+         return m_context;
+         }
+
+      const T& get() const
+         {
+         return m_context;
+         }
+
+      T operator->()
+         {
+         return m_context;
+         }
+
+   private:
+      template<class T>
+      void close()
+         {
+         static_assert(false, "Handle_Guard is not available for this type");
+         }
+
+      template<>
+      void close<PCCERT_CONTEXT> ()
+         {
+         if(m_context)
+            {
+            CertFreeCertificateContext(m_context);
+            }
+         }
+
+      template<>
+      void close<HCERTSTORE> ()
+         {
+         if(m_context)
+            {
+            CertCloseStore(m_context, 0);
+            }
+         }
+
+      T m_context;
+   };
+}
+
+namespace Botan {
+Certificate_Store_Windows::Certificate_Store_Windows() {}
+
+std::vector<X509_DN> Certificate_Store_Windows::all_subjects() const
+   {
+   std::vector<X509_DN> subject_dns;
+   for(auto& store_name : ::cert_store_names)
+      {
+      Handle_Guard<HCERTSTORE> windows_cert_store = openCertStore(store_name.c_str());
+      Handle_Guard<PCCERT_CONTEXT> cert_context = nullptr;
+
+      // Handle_Guard::assign exchanges the underlying pointer. No RAII is needed here, because the Windows API takes care of
+      // freeing the previous context.
+      while(cert_context.assign(CertEnumCertificatesInStore(windows_cert_store.get(), cert_context.get())))
+         {
+         X509_Certificate cert(cert_context->pbCertEncoded, cert_context->cbCertEncoded);
+         subject_dns.push_back(cert.subject_dn());
+         }
+      }
+
+   return subject_dns;
+   }
+
+std::shared_ptr<const X509_Certificate>
+Certificate_Store_Windows::find_cert(const Botan::X509_DN&           subject_dn,
+                                     const std::vector<uint8_t>& key_id) const
+   {
+   const auto certs = find_all_certs(subject_dn, key_id);
+   return certs.empty() ? nullptr : certs.front();
+   }
+
+std::vector<std::shared_ptr<const X509_Certificate>> Certificate_Store_Windows::find_all_certs(
+         const X509_DN& subject_dn,
+         const std::vector<uint8_t>& key_id) const
+   {
+   std::vector<uint8_t> dn_data;
+   DER_Encoder encoder(dn_data);
+   subject_dn.encode_into(encoder);
+
+   CERT_NAME_BLOB blob;
+   blob.cbData = static_cast<DWORD>(dn_data.size());
+   blob.pbData = reinterpret_cast<BYTE*>(dn_data.data());
+
+   std::vector<std::shared_ptr<const X509_Certificate>> certs;
+   for(auto& store_name : ::cert_store_names)
```

Comment:
If key_id is set it seems better to search by key id since that should be unique or nearly so, which will reduce the number of certs that must be parsed.